File: 0065-Valid Number/0065-Valid Number.cpp
class Solution {
public:
    bool isNumber(string s) {
        int n = s.size();        
        int start = 0, end = n - 1;
        while (start <= end && s[start] == ' ') {
            ++start;
        }
        
        if (start > end) {
            return false;
        }
        
        while (end >= start && s[end] == ' ') {
            --end;
        }
        
        if (s[start] == '+' || s[start] == '-') {
            ++start;
        }
        
        bool num = false, dot = false, exp = false;
        while (start <= end) {
            char c = s[start];
            if (isdigit(c)) {
                num = true;
            }
            else if (c == '.') {
                if (dot || exp) {
                    return false;
                }
                dot = true;
            }
            else if (c == 'e') {
                if (exp || !num) {
                    return false;
                }
                exp = true;
                num = false;
            }
            else if (c == '+' || c == '-') {
                if (s[start - 1] != 'e') {
                    return false;
                }
            }
            else {
                return false;
            }
            ++start;
        }
        
        return num;
    }
};

File: 0001-0100/0065-Valid Number/0065-Valid Number.py
class Solution:
    def isNumber(self, s: str) -> bool:
        start = 0
        end = len(s) - 1
        while start <= end and s[start] == ' ':
            start += 1
        if start > end:
            return False
        while start <= end and s[end] == ' ':
            end -= 1
        if s[start] == '+' or s[start] == '-':
            start += 1
        num = dot = exp = False
        while start <= end:
            if s[start].isdigit():
                num = True
            elif s[start] == '.':
                if dot or exp:
                    return False
                dot = True
            elif s[start] == 'e':
                if not num or exp:
                    return False
                exp = True
                num = False
            elif s[start] == '+' or s[start] == '-':
                if s[start - 1] != 'e':
                    return False
            else:
                return False
            start += 1
        return num
